lib/Stan_math_code_gen.ml

```diff
@@ -191,35 +194,31 @@ let pp_statement pp_statement_with_meta ppf s =
             pf ppf ",@ %s" extra_arg )
         , (fdargs, "std::ostream* pstream__") ) ;
       pf ppf " " ;
-      pp_block ppf
-        ( (fun ppf fdbody ->
-            let text = pf ppf "%s@;" in
-            pf ppf
-              "@[<hv 8>typedef typename \
-               boost::math::tools::promote_args<%a>::type \
-               local_scalar_t__;@]@ "
-              (list ~sep:comma string) argtypetemplates ;
-            text "typedef local_scalar_t__ fun_return_scalar_t__;" ;
-            text "const static bool propto__ = true;" ;
-            text "(void) propto__;" ;
-            text
-              "local_scalar_t__ \
-               DUMMY_VAR__(std::numeric_limits<double>::quiet_NaN());" ;
-            text "(void) DUMMY_VAR__;  // suppress unused var warning" ;
-            text "int current_statement_begin__ = -1;" ;
-            pp_located_error ppf (pp_statement_with_meta, fdbody, None) )
-        , fdbody )
-
-let rec pp_statement_loc ppf {sloc; stmt} =
-  (*
-  Sexp.to_string_hum [%sexp (stmt : stmt_loc statement)] |> print_endline ;
-*)
-  pf ppf "current_statement_loc__ = \"%s\";@;" sloc ;
-  pp_statement pp_statement_loc ppf stmt
+      match fdbody.stmt with
+      | Skip -> pf ppf ";@ "
```

Comment:
This right here is the reason I had to convert pp_statement from taking in an `'a statement` to knowing exactly what `'a` would be. Needed to dissect `fdbody` to see if it was just a Skip and not emit code, and we still don't care about the metadata in this function but apparently the typechecker needs us to know what it will be so it can inspect. Not sure if @VMatthijs can think of any ways around this, it probably eventually reduces to needing row polymorphism. I tried to hack it by just pretty printing the body and checking if it's empty, but unfortunately the locations will always be printed out even if the statement is a Skip (which, actually thinking about it now, maybe that shouldn't be true). 

Relatedly, I'm wondering if it makes sense to just put the locations in the MIR as a variant of statement. The trouble I perceive would be that it could make inter-statement dependency tracking more difficult than just checking where Vars are used and where they come from or are modified. I'd love it if anyone could chime in here and hash this out with me now, or maybe we should just try it for now.

Reply:
I may need a face-to-face explanation of the problem.

Reply:
Sounds good. Are you free tomorrow afternoon?